### PR TITLE
Framework: Upgrade to React 15.1.0

### PR DESCRIPTION
**Pull request by ockham**
Upgrade Calypso to React 15.1.0, and its peer dependencies to compatible versions (`qrcode.react` to 0.6.1, and `react-tap-event-plugin` to 1.0.0).

* React blog entry with Changelog at https://facebook.github.io/react/blog/2016/04/07/react-v15.html
* `qrcode.react` 0.6.1 Changelog: https://github.com/zpao/qrcode.react/blob/master/CHANGELOG.md#061---2016-04-13
* (No changelog available for `react-tap-event-plugin`)

This is described as an upgrade with less impact than previous ones (like 0.13 -> 0.14). However, it's quite possible that I've missed subtle bugs, so please be sure to thorougly test the sections you're familiar with.

Most notably, React 15 [removes implicit `<span>`s](https://facebook.github.io/react/blog/2016/04/07/react-v15.html#no-more-extra-ltspangts) previously inserted around text nodes (considering them an implementation detail). Unfortunately, some of our styling relied on those `<span>`s. I tried to track them down (aided by visual regressions thanks to Flow Patrol, see p6jOYE-sc-p2). However, it is quite likely that I missed some of them, so I'd ask reviewers to watch out for styling oddities in those sections you're familiar with.

Furthermore, React 15 [enables all SVG attributes](https://facebook.github.io/react/blog/2016/04/07/react-v15.html#improved-svg-support) -- but as with all things React, they're `camelCased`. We do have some inline `<svg>`s, but some of them have been using attributes like `fill-rule` that have been ignored. We should change those to camelCase. This will be more complicated with generated code (`Gridicons`, `SocialLogos`). Fortunately, AFAICS this doesn't seem to break any existing functionality, so we tackle this separately (#5950).

Deprecations:

* `LinkedStateMixin` and `valueLink` are now [deprecated](https://facebook.github.io/react/blog/2016/04/07/react-v15.html#new-deprecations-introduced-with-a-warning) due to very low popularity, so we should change our components to no longer rely on them (#5951).

* React 15 warns you on a `<input value={ null }>` and offers you to clarify your intention. To fix the warning, you may explicitly pass an empty string to clear a controlled input, or pass `undefined` to make the input uncontrolled.

---

Note that the gruntwork for this PR was actually done in a number of separate, preparatory PRs (see linked merged PRs below). Some internal (undocumented) features our tests were relying on (`__reactAutoBindMap`!) were removed, so I fixed them beforehand to no longer rely on those features but only on documented ones.

**Verified Working**
 - [x] Importer (@dmsnell)
 - [x] SEO settings (@dmsnell)

Test live: https://calypso.live/?branch=update/react-15-0-1

### CODE REVIEW

**Comment by aduth** [comment]
If SVG changes are an optional enhancement, should we wait and make those changes after bumping the version, so as to reduce the likelihood of breakage?

**Comment by ockham** [comment]
> If SVG changes are an optional enhancement, should we wait and make those changes after bumping the version, so as to reduce the likelihood of breakage?

Yeah, good idea. I branched off #5164 with the SVG inlining.

**Comment by ockham** [comment]
Okay, I think I've been staring at this alone long enough :-D Setting Status to _Needs Review_.

File: npm-shrinkwrap.json [hunk 1]
```diff
@@ -2777,33 +2794,36 @@
...
       "dependencies": {
         "core-js": {
           "version": "1.2.6"
         },
         "fbjs": {
-          "version": "0.3.2"
+          "version": "0.8.3"
+        },
+        "immutable": {
+          "version": "3.8.1"
```
Comment (blowery) [inline npm-shrinkwrap.json]: maybe we should upgrade our immutable too? we're on 3.7.5

File: npm-shrinkwrap.json [hunk 2]
```diff
@@ -1738,7 +1738,24 @@
       }
     },
     "interpolate-components": {
-      "version": "1.0.4"
+      "version": "1.0.4",
```
Comment (blowery) [inline npm-shrinkwrap.json]: this is ours, we should probably update it as well

Comment (ockham) [inline npm-shrinkwrap.json]: https://github.com/Automattic/interpolate-components/pull/1 :-)

File: npm-shrinkwrap.json (hunk 1 above)
Comment (ockham) [inline npm-shrinkwrap.json]: #5958

**Comment by aduth** [comment]
Doing a bit of regex scanning for SVG attributes potentially needing updates:

- `enable-background` to `enableBackground` ([reference](https://facebook.github.io/react/docs/tags-and-attributes.html#svg-attributes))
 - https://github.com/Automattic/wp-calypso/blob/de5eba6/client/my-sites/plan-features/header.jsx#L95
 - https://github.com/Automattic/wp-calypso/blob/de5eba6/client/my-sites/plan-features/header.jsx#L297
 - https://github.com/Automattic/wp-calypso/blob/de5eba6/client/reader/post-excerpt-link/index.jsx#L51

**Comment by aduth** [comment]
Tried to find missed `span`s and came up short, aside from one in DevDocs:

- https://github.com/Automattic/wp-calypso/blob/de5eba6/client/components/global-notices/docs/example.jsx#L53

**Comment by ockham** [comment]
> Tried to find missed `span`s and came up short, aside from one in DevDocs

Great spot! Fixed in 0325d00.

**Comment by ockham** [comment]
Resisted the temptation for now. `enableBackground` fix in ba0dd2c.

**Comment by aduth** [comment]
One more:

- https://github.com/Automattic/wp-calypso/blob/de5eba6/client/my-sites/upgrades/checkout/privacy-protection-example.jsx ([styles](https://github.com/Automattic/wp-calypso/blob/de5eba686ff440834462a86c76c342b639086a5e/client/my-sites/upgrades/checkout/style.scss#L755-L761))

Potentially unreached/unused style block (cc @jancavan @blowery ):

- https://github.com/Automattic/wp-calypso/blob/de5eba686ff440834462a86c76c342b639086a5e/client/reader/_style.scss#L134-L151

**Comment by ockham** [comment]
> Potentially unreached/unused style block (cc @jancavan @blowery ):

> - https://github.com/Automattic/wp-calypso/blob/de5eba686ff440834462a86c76c342b639086a5e/client/reader/_style.scss#L134-L151

I came across that one, and I think that class is probably provided as part of some pre-rendered markup by an endpoint (reader related?).

**Comment by ockham** [comment]
One more:

- https://github.com/Automattic/wp-calypso/blob/de5eba6/client/my-sites/upgrades/checkout/privacy-protection-example.jsx ([styles](https://github.com/Automattic/wp-calypso/blob/de5eba686ff440834462a86c76c342b639086a5e/client/my-sites/upgrades/checkout/style.scss#L755-L761))

4794ec7

**Comment by blowery** [comment]
> I came across that one, and I think that class is probably provided as part of some pre-rendered markup by an endpoint (reader related?).

Yup, generated by the API / Reader API for reblogs.

File: npm-shrinkwrap.json (hunk 2 above)
Comment (blowery) [inline npm-shrinkwrap.json]: Need to bump our package.json and re-gen shrinkwrap now that 1.0.5 is out

File: npm-shrinkwrap.json (hunk 1 above)
Comment (blowery) [inline npm-shrinkwrap.json]: need to reshrinkwrap now that we bumped immutable

**Comment by blowery** [comment]
Can we rebase this and re-shrinkwrap it now that immutable has been bumped to match react 15 and interpolate-components@1.0.5 is on react 15 too?

File: npm-shrinkwrap.json (hunk 2 above)
Comment (ockham) [inline npm-shrinkwrap.json]: Needs `i18n-calypso` to be bumped first -- https://github.com/Automattic/i18n-calypso/issues/3

**Comment by ockham** [comment]
@blowery Rebased, reshrunk.

**Comment by blowery** [comment]
@ockham woo! shinkwrap looks so much better!

**Comment by gwwar** [comment]
👍 Looking great!

**Comment by blowery** [comment]
Tested reader, seems to work as expected

**Comment by ockham** [comment]
Thanks for reviewing @aduth @blowery @gwwar!

**Pull request merged**